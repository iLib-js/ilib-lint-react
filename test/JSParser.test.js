/*
 * JSParser.test.js - test the React JS parser
 *
 * Copyright Â© 2023 Box, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { ResourceString } from 'ilib-tools-common';

import JSParser from '../src/parsers/JSParser.js';

import { Result, IntermediateRepresentation } from 'i18nlint-common';

describe("testJSParser", () => {
    test("JSParserConstructorEmpty", () => {
        expect.assertions(1);

        const parser = new JSParser();
        expect(parser).toBeTruthy();
    });

    test("JSParserConstructorPath", () => {
        expect.assertions(1);

        const parser = new JSParser({
            filePath: "./test/testfiles/testfile.js"
        });
        expect(parser).toBeTruthy();
    });

    test("JSParserGetDescription", () => {
        expect.assertions(2);

        const parser = new JSParser();
        expect(parser).toBeTruthy();

        expect(parser.getDescription()).toBe("A parser for JS files.");
    });

    test("JSParserGetName", () => {
        expect.assertions(2);

        const parser = new JSParser();
        expect(parser).toBeTruthy();

        expect(parser.getName()).toBe("js");
    });

    test("JSParserGetExtensions", () => {
        expect.assertions(2);

        const parser = new JSParser();
        expect(parser).toBeTruthy();

        expect(parser.getExtensions()).toStrictEqual([ "js" ]);
    });

    test("JSParserSimple", () => {
        expect.assertions(3);

        const parser = new JSParser();
        expect(parser).toBeTruthy();

        const actual = parser.parseString("import foo from '../src/index.js';", "x/y");
        expect(actual).toBeTruthy();
        const actualSimplified = JSON.parse(JSON.stringify(actual));

        const expected = {
            "type": "ast-jstree",
            "ir": {
                "type": 'Program',
                "start": 0,
                "end": 34,
                "loc": {
                    "start": { "line": 1, "column": 0 },
                    "end": { "line": 1, "column": 34 }
                },
                "body": [
                    {
                        "type": 'ImportDeclaration',
                        "start": 0,
                        "end": 34,
                        "loc": {
                            "start": { "line": 1, "column": 0 },
                            "end": { "line": 1, "column": 34 }
                        },
                        "specifiers": [
                            {
                                "type": 'ImportDefaultSpecifier',
                                "start": 7,
                                "end": 10,
                                "loc": {
                                    "start": { "line": 1, "column": 7 },
                                    "end": { "line": 1, "column": 10 }
                                },
                                "local": {
                                    "type": 'Identifier',
                                    "start": 7,
                                    "end": 10,
                                    "loc": {
                                        "start": { "line": 1, "column": 7 },
                                        "end": { "line": 1, "column": 10 }
                                    },
                                    "name": 'foo'
                                }
                            }
                        ],
                        "source": {
                            "type": 'Literal',
                            "start": 16,
                            "end": 33,
                            "loc": {
                                "start": { "line": 1, "column": 16 },
                                "end": { "line": 1, "column": 33 }
                            },
                            "value": '../src/index.js',
                            "raw": "'../src/index.js'"
                        }
                    }
                ],
                "sourceType": 'module'
            },
            filePath: "x/y"
        };
        expect(actualSimplified).toStrictEqual(expected);
    });

    test("JSParserMoreComplex", () => {
        expect.assertions(3);

        const parser = new JSParser();
        expect(parser).toBeTruthy();

        const actual = parser.parseString(
            `// comment
            import foo from '../src/index.js';

            const str = "String";
            `, "x/y");
        expect(actual).toBeTruthy();
        const actualSimplified = JSON.parse(JSON.stringify(actual));

        const expected = {
            "type": "ast-jstree",
            "ir": {
                "type": "Program",
                "start": 0,
                "end": 105,
                "loc": {
                    "start": {
                        "line": 1,
                        "column": 0
                    },
                    "end": {
                        "line": 5,
                        "column": 12
                    }
                },
                "body": [
                    {
                        "type": "ImportDeclaration",
                        "start": 23,
                        "end": 57,
                        "loc": {
                            "start": {
                                "line": 2,
                                "column": 12
                            },
                            "end": {
                                "line": 2,
                                "column": 46
                            }
                        },
                        "specifiers": [
                            {
                                "type": "ImportDefaultSpecifier",
                                "start": 30,
                                "end": 33,
                                "loc": {
                                    "start": {
                                        "line": 2,
                                        "column": 19
                                    },
                                    "end": {
                                        "line": 2,
                                        "column": 22
                                    }
                                },
                                "local": {
                                    "type": "Identifier",
                                    "start": 30,
                                    "end": 33,
                                    "loc": {
                                        "start": {
                                            "line": 2,
                                            "column": 19
                                        },
                                        "end": {
                                            "line": 2,
                                            "column": 22
                                        }
                                    },
                                    "name": "foo"
                                }
                            }
                        ],
                        "source": {
                            "type": "Literal",
                            "start": 39,
                            "end": 56,
                            "loc": {
                                "start": {
                                    "line": 2,
                                    "column": 28
                                },
                                "end": {
                                    "line": 2,
                                    "column": 45
                                }
                            },
                            "value": "../src/index.js",
                            "raw": "'../src/index.js'"
                        }
                    },
                    {
                        "type": "VariableDeclaration",
                        "start": 71,
                        "end": 92,
                        "loc": {
                            "start": {
                                "line": 4,
                                "column": 12
                            },
                            "end": {
                                "line": 4,
                                "column": 33
                            }
                        },
                        "declarations": [
                            {
                                "type": "VariableDeclarator",
                                "start": 77,
                                "end": 91,
                                "loc": {
                                    "start": {
                                        "line": 4,
                                        "column": 18
                                    },
                                    "end": {
                                        "line": 4,
                                        "column": 32
                                    }
                                },
                                "id": {
                                    "type": "Identifier",
                                    "start": 77,
                                    "end": 80,
                                    "loc": {
                                        "start": {
                                            "line": 4,
                                            "column": 18
                                        },
                                        "end": {
                                            "line": 4,
                                            "column": 21
                                        }
                                    },
                                    "name": "str"
                                },
                                "init": {
                                    "type": "Literal",
                                    "start": 83,
                                    "end": 91,
                                    "loc": {
                                        "start": {
                                            "line": 4,
                                            "column": 24
                                        },
                                        "end": {
                                            "line": 4,
                                            "column": 32
                                        }
                                    },
                                    "value": "String",
                                    "raw": "\"String\""
                                }
                            }
                        ],
                        "kind": "const"
                    }
                ],
                "sourceType": "module"
            },
            "filePath": "x/y"
        };
        expect(actualSimplified).toStrictEqual(expected);
    });
});

