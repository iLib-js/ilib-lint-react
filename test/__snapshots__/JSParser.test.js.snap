// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testJSParser JS parse of real file 1`] = `
[
  IntermediateRepresentation {
    "filePath": "./test/testfiles/testfile2.js",
    "ir": Node {
      "body": [
        Node {
          "end": 31,
          "loc": SourceLocation {
            "end": Position {
              "column": 31,
              "line": 1,
            },
            "start": Position {
              "column": 0,
              "line": 1,
            },
          },
          "source": Node {
            "end": 30,
            "loc": SourceLocation {
              "end": Position {
                "column": 30,
                "line": 1,
              },
              "start": Position {
                "column": 23,
                "line": 1,
              },
            },
            "raw": "'react'",
            "start": 23,
            "type": "Literal",
            "value": "react",
          },
          "specifiers": [
            Node {
              "end": 17,
              "loc": SourceLocation {
                "end": Position {
                  "column": 17,
                  "line": 1,
                },
                "start": Position {
                  "column": 7,
                  "line": 1,
                },
              },
              "local": Node {
                "end": 17,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 17,
                    "line": 1,
                  },
                  "start": Position {
                    "column": 12,
                    "line": 1,
                  },
                },
                "name": "React",
                "start": 12,
                "type": "Identifier",
              },
              "start": 7,
              "type": "ImportNamespaceSpecifier",
            },
          ],
          "start": 0,
          "type": "ImportDeclaration",
        },
        Node {
          "body": Node {
            "body": [
              Node {
                "computed": false,
                "end": 138,
                "key": Node {
                  "end": 91,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 19,
                      "line": 4,
                    },
                    "start": Position {
                      "column": 4,
                      "line": 4,
                    },
                  },
                  "name": "getSourceString",
                  "start": 76,
                  "type": "Identifier",
                },
                "kind": "method",
                "loc": SourceLocation {
                  "end": Position {
                    "column": 5,
                    "line": 6,
                  },
                  "start": Position {
                    "column": 4,
                    "line": 4,
                  },
                },
                "start": 76,
                "static": false,
                "type": "MethodDefinition",
                "value": Node {
                  "async": false,
                  "body": Node {
                    "body": [
                      Node {
                        "argument": Node {
                          "computed": false,
                          "end": 131,
                          "loc": SourceLocation {
                            "end": Position {
                              "column": 35,
                              "line": 5,
                            },
                            "start": Position {
                              "column": 15,
                              "line": 5,
                            },
                          },
                          "object": Node {
                            "end": 115,
                            "loc": SourceLocation {
                              "end": Position {
                                "column": 19,
                                "line": 5,
                              },
                              "start": Position {
                                "column": 15,
                                "line": 5,
                              },
                            },
                            "start": 111,
                            "type": "ThisExpression",
                          },
                          "optional": false,
                          "property": Node {
                            "end": 131,
                            "loc": SourceLocation {
                              "end": Position {
                                "column": 35,
                                "line": 5,
                              },
                              "start": Position {
                                "column": 20,
                                "line": 5,
                              },
                            },
                            "name": "getSourceString",
                            "start": 116,
                            "type": "Identifier",
                          },
                          "start": 111,
                          "type": "MemberExpression",
                        },
                        "end": 132,
                        "loc": SourceLocation {
                          "end": Position {
                            "column": 36,
                            "line": 5,
                          },
                          "start": Position {
                            "column": 8,
                            "line": 5,
                          },
                        },
                        "start": 104,
                        "type": "ReturnStatement",
                      },
                    ],
                    "end": 138,
                    "loc": SourceLocation {
                      "end": Position {
                        "column": 5,
                        "line": 6,
                      },
                      "start": Position {
                        "column": 22,
                        "line": 4,
                      },
                    },
                    "start": 94,
                    "type": "BlockStatement",
                  },
                  "end": 138,
                  "expression": false,
                  "generator": false,
                  "id": null,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 5,
                      "line": 6,
                    },
                    "start": Position {
                      "column": 19,
                      "line": 4,
                    },
                  },
                  "params": [],
                  "start": 91,
                  "type": "FunctionExpression",
                },
              },
              Node {
                "computed": false,
                "end": 196,
                "key": Node {
                  "end": 150,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 10,
                      "line": 8,
                    },
                    "start": Position {
                      "column": 4,
                      "line": 8,
                    },
                  },
                  "name": "render",
                  "start": 144,
                  "type": "Identifier",
                },
                "kind": "method",
                "loc": SourceLocation {
                  "end": Position {
                    "column": 5,
                    "line": 10,
                  },
                  "start": Position {
                    "column": 4,
                    "line": 8,
                  },
                },
                "start": 144,
                "static": false,
                "type": "MethodDefinition",
                "value": Node {
                  "async": false,
                  "body": Node {
                    "body": [
                      Node {
                        "argument": Node {
                          "computed": false,
                          "end": 189,
                          "loc": SourceLocation {
                            "end": Position {
                              "column": 34,
                              "line": 9,
                            },
                            "start": Position {
                              "column": 15,
                              "line": 9,
                            },
                          },
                          "object": Node {
                            "computed": false,
                            "end": 180,
                            "loc": SourceLocation {
                              "end": Position {
                                "column": 25,
                                "line": 9,
                              },
                              "start": Position {
                                "column": 15,
                                "line": 9,
                              },
                            },
                            "object": Node {
                              "end": 174,
                              "loc": SourceLocation {
                                "end": Position {
                                  "column": 19,
                                  "line": 9,
                                },
                                "start": Position {
                                  "column": 15,
                                  "line": 9,
                                },
                              },
                              "start": 170,
                              "type": "ThisExpression",
                            },
                            "optional": false,
                            "property": Node {
                              "end": 180,
                              "loc": SourceLocation {
                                "end": Position {
                                  "column": 25,
                                  "line": 9,
                                },
                                "start": Position {
                                  "column": 20,
                                  "line": 9,
                                },
                              },
                              "name": "props",
                              "start": 175,
                              "type": "Identifier",
                            },
                            "start": 170,
                            "type": "MemberExpression",
                          },
                          "optional": false,
                          "property": Node {
                            "end": 189,
                            "loc": SourceLocation {
                              "end": Position {
                                "column": 34,
                                "line": 9,
                              },
                              "start": Position {
                                "column": 26,
                                "line": 9,
                              },
                            },
                            "name": "children",
                            "start": 181,
                            "type": "Identifier",
                          },
                          "start": 170,
                          "type": "MemberExpression",
                        },
                        "end": 190,
                        "loc": SourceLocation {
                          "end": Position {
                            "column": 35,
                            "line": 9,
                          },
                          "start": Position {
                            "column": 8,
                            "line": 9,
                          },
                        },
                        "start": 163,
                        "type": "ReturnStatement",
                      },
                    ],
                    "end": 196,
                    "loc": SourceLocation {
                      "end": Position {
                        "column": 5,
                        "line": 10,
                      },
                      "start": Position {
                        "column": 13,
                        "line": 8,
                      },
                    },
                    "start": 153,
                    "type": "BlockStatement",
                  },
                  "end": 196,
                  "expression": false,
                  "generator": false,
                  "id": null,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 5,
                      "line": 10,
                    },
                    "start": Position {
                      "column": 10,
                      "line": 8,
                    },
                  },
                  "params": [],
                  "start": 150,
                  "type": "FunctionExpression",
                },
              },
            ],
            "end": 198,
            "loc": SourceLocation {
              "end": Position {
                "column": 1,
                "line": 11,
              },
              "start": Position {
                "column": 37,
                "line": 3,
              },
            },
            "start": 70,
            "type": "ClassBody",
          },
          "end": 198,
          "id": Node {
            "end": 45,
            "loc": SourceLocation {
              "end": Position {
                "column": 12,
                "line": 3,
              },
              "start": Position {
                "column": 6,
                "line": 3,
              },
            },
            "name": "Plural",
            "start": 39,
            "type": "Identifier",
          },
          "loc": SourceLocation {
            "end": Position {
              "column": 1,
              "line": 11,
            },
            "start": Position {
              "column": 0,
              "line": 3,
            },
          },
          "start": 33,
          "superClass": Node {
            "computed": false,
            "end": 69,
            "loc": SourceLocation {
              "end": Position {
                "column": 36,
                "line": 3,
              },
              "start": Position {
                "column": 21,
                "line": 3,
              },
            },
            "object": Node {
              "end": 59,
              "loc": SourceLocation {
                "end": Position {
                  "column": 26,
                  "line": 3,
                },
                "start": Position {
                  "column": 21,
                  "line": 3,
                },
              },
              "name": "React",
              "start": 54,
              "type": "Identifier",
            },
            "optional": false,
            "property": Node {
              "end": 69,
              "loc": SourceLocation {
                "end": Position {
                  "column": 36,
                  "line": 3,
                },
                "start": Position {
                  "column": 27,
                  "line": 3,
                },
              },
              "name": "Component",
              "start": 60,
              "type": "Identifier",
            },
            "start": 54,
            "type": "MemberExpression",
          },
          "type": "ClassDeclaration",
        },
        Node {
          "declaration": Node {
            "end": 221,
            "loc": SourceLocation {
              "end": Position {
                "column": 21,
                "line": 13,
              },
              "start": Position {
                "column": 15,
                "line": 13,
              },
            },
            "name": "Plural",
            "start": 215,
            "type": "Identifier",
          },
          "end": 222,
          "loc": SourceLocation {
            "end": Position {
              "column": 22,
              "line": 13,
            },
            "start": Position {
              "column": 0,
              "line": 13,
            },
          },
          "start": 200,
          "type": "ExportDefaultDeclaration",
        },
      ],
      "end": 222,
      "loc": SourceLocation {
        "end": Position {
          "column": 22,
          "line": 13,
        },
        "start": Position {
          "column": 0,
          "line": 1,
        },
      },
      "sourceType": "module",
      "start": 0,
      "type": "Program",
    },
    "stats": undefined,
    "type": "ast-jstree",
  },
  IntermediateRepresentation {
    "filePath": "./test/testfiles/testfile2.js",
    "ir": "import * as React from 'react';

class Plural extends React.Component {
    getSourceString() {
        return this.getSourceString;
    }

    render() {
        return this.props.children;
    }
}

export default Plural;",
    "stats": undefined,
    "type": "string",
  },
]
`;

exports[`testJSParser JSParserMoreComplex 1`] = `
{
  "filePath": "x/y",
  "ir": {
    "body": [
      {
        "end": 57,
        "loc": {
          "end": {
            "column": 46,
            "line": 2,
          },
          "start": {
            "column": 12,
            "line": 2,
          },
        },
        "source": {
          "end": 56,
          "loc": {
            "end": {
              "column": 45,
              "line": 2,
            },
            "start": {
              "column": 28,
              "line": 2,
            },
          },
          "raw": "'../src/index.js'",
          "start": 39,
          "type": "Literal",
          "value": "../src/index.js",
        },
        "specifiers": [
          {
            "end": 33,
            "loc": {
              "end": {
                "column": 22,
                "line": 2,
              },
              "start": {
                "column": 19,
                "line": 2,
              },
            },
            "local": {
              "end": 33,
              "loc": {
                "end": {
                  "column": 22,
                  "line": 2,
                },
                "start": {
                  "column": 19,
                  "line": 2,
                },
              },
              "name": "foo",
              "start": 30,
              "type": "Identifier",
            },
            "start": 30,
            "type": "ImportDefaultSpecifier",
          },
        ],
        "start": 23,
        "type": "ImportDeclaration",
      },
      {
        "declarations": [
          {
            "end": 91,
            "id": {
              "end": 80,
              "loc": {
                "end": {
                  "column": 21,
                  "line": 4,
                },
                "start": {
                  "column": 18,
                  "line": 4,
                },
              },
              "name": "str",
              "start": 77,
              "type": "Identifier",
            },
            "init": {
              "end": 91,
              "loc": {
                "end": {
                  "column": 32,
                  "line": 4,
                },
                "start": {
                  "column": 24,
                  "line": 4,
                },
              },
              "raw": ""String"",
              "start": 83,
              "type": "Literal",
              "value": "String",
            },
            "loc": {
              "end": {
                "column": 32,
                "line": 4,
              },
              "start": {
                "column": 18,
                "line": 4,
              },
            },
            "start": 77,
            "type": "VariableDeclarator",
          },
        ],
        "end": 92,
        "kind": "const",
        "loc": {
          "end": {
            "column": 33,
            "line": 4,
          },
          "start": {
            "column": 12,
            "line": 4,
          },
        },
        "start": 71,
        "type": "VariableDeclaration",
      },
    ],
    "end": 105,
    "loc": {
      "end": {
        "column": 12,
        "line": 5,
      },
      "start": {
        "column": 0,
        "line": 1,
      },
    },
    "sourceType": "module",
    "start": 0,
    "type": "Program",
  },
  "type": "ast-jstree",
}
`;

exports[`testJSParser JSParserSimple 1`] = `
{
  "filePath": "x/y",
  "ir": {
    "body": [
      {
        "end": 34,
        "loc": {
          "end": {
            "column": 34,
            "line": 1,
          },
          "start": {
            "column": 0,
            "line": 1,
          },
        },
        "source": {
          "end": 33,
          "loc": {
            "end": {
              "column": 33,
              "line": 1,
            },
            "start": {
              "column": 16,
              "line": 1,
            },
          },
          "raw": "'../src/index.js'",
          "start": 16,
          "type": "Literal",
          "value": "../src/index.js",
        },
        "specifiers": [
          {
            "end": 10,
            "loc": {
              "end": {
                "column": 10,
                "line": 1,
              },
              "start": {
                "column": 7,
                "line": 1,
              },
            },
            "local": {
              "end": 10,
              "loc": {
                "end": {
                  "column": 10,
                  "line": 1,
                },
                "start": {
                  "column": 7,
                  "line": 1,
                },
              },
              "name": "foo",
              "start": 7,
              "type": "Identifier",
            },
            "start": 7,
            "type": "ImportDefaultSpecifier",
          },
        ],
        "start": 0,
        "type": "ImportDeclaration",
      },
    ],
    "end": 34,
    "loc": {
      "end": {
        "column": 34,
        "line": 1,
      },
      "start": {
        "column": 0,
        "line": 1,
      },
    },
    "sourceType": "module",
    "start": 0,
    "type": "Program",
  },
  "type": "ast-jstree",
}
`;
